category_theory/limits/limits.lean

```diff
@@ -0,0 +1,584 @@
+-- Copyright (c) 2018 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison, Reid Barton, Mario Carneiro
+
+import category_theory.whiskering
+import category_theory.yoneda
+import category_theory.limits.cones
+
+open category_theory category_theory.category category_theory.functor
+
+namespace category_theory.limits
+
+universes u u' u'' v w
+
+variables {J : Type v} [small_category J]
+variables {C : Type u} [𝒞 : category.{u v} C]
+include 𝒞
+
+section is_limit
+variables {F : J ⥤ C}
+
+/-- A cone `t` on `F` is a limit cone if each cone on `F` admits a unique
+  cone morphism to `t`. -/
+structure is_limit (t : cone F) :=
+(lift  : Π (s : cone F), s.X ⟶ t.X)
+(fac'  : ∀ (s : cone F) (j : J), lift s ≫ t.π.app j = s.π.app j . obviously)
+(uniq' : ∀ (s : cone F) (m : s.X ⟶ t.X) (w : ∀ j : J, m ≫ t.π.app j = s.π.app j),
+  m = lift s . obviously)
+
+restate_axiom is_limit.fac'
+attribute [simp] is_limit.fac
+restate_axiom is_limit.uniq'
+
+instance is_limit_subsingleton {t : cone F} : subsingleton (is_limit t) :=
+⟨by intros P Q; cases P; cases Q; congr; ext; solve_by_elim⟩
+
```

Comment:
I guess it would make sense to start a `is_limit` namespace around here. The following 10 (circa) lemmas all start with `is_limit.`.

Reply:
Okay, I've added namespaces for these.